numpy/core/src/multiarray/item_selection.c

```diff
@@ -2096,10 +2103,13 @@ PyArray_CountNonzero(PyArrayObject *self)
             if (nonzero(data, self)) {
                 ++nonzero_count;
             }
+            if (needs_api && PyErr_Occurred()) {
+                return -1;
+            }
             data += stride;
         }
 
-        return nonzero_count;
+        return PyErr_Occurred() ? -1 : nonzero_count;
```

Comment:
I  half expected that cast, but it is silly, because python convention is to return an error indictor value, and that is also the case here (and nonzero is not much more then that function):

> int PyObject_IsTrue(PyObject *o)
> Returns 1 if the object o is considered to be true, and 0 otherwise. This is equivalent to the Python expression not not o. On failure, return -1.

Ignoring that means you do a more expensive check far more often. @ahaldane of course you need to break the inner loop, once the python error is set, you need to either clear it or stop doing python stuff, continuing the loop would do python stuff.

About "in the future we might to catch something", well, its a pretty safe thing to do, though I doubt there is much use here. If you have a type that can easily throw errors, it will "need_api". Though yes, we have a bit of problems with indicating errors for types that don't typically give errors....